proposals/0046-astro-env.md

```diff
@@ -0,0 +1,573 @@
+// ...
+
+export const createExports = (manifest: SSRManifest) => {
+  const app = new App(manifest);
+  app.setGetEnv(getEnv);
+  // ...
+};
+```
+
+This sets `getEnv` globally but some adapters need to set it per request. Well, you can call this method whenever you want so it can be right before `app.render`:
+
+```ts
+// ...
+
+export const createExports = (manifest: SSRManifest) => {
+  const app = new App(manifest);
+
+  const fetch = async (
+    request: Request & CLOUDFLARE_REQUEST,
+    env: Env,
+    context: ExecutionContext
+  ) => {
+    // ...
+    app.setGetEnv((key) => env[key]);
+    const response = await app.render(request, { routeData, locals });
+    // ...
+  };
+  // ...
+};
+```
+
+##### Handling variables in dev and build
+
+Since in dev and build variables are taken from `process.env`, if env is coming from another source (eg. Cloudflare `platformProxy`) then it's important to call `overrideProcessEnv`:
+
+```ts
+import { overrideProcessEnv } from "astro/runtime/server/astro-env.js";
+import { getEnv } from "./utils.js";
+
+const integration = (): AstroIntegration => {
+  let config;
+
+  return {
+    name: "adapter",
+    hooks: {
+      "astro:config:done": (params) => {
+        config = params.config;
+      },
+      // Other hooks can be used
+      "astro:server:setup": () => {
+        const { env } = platformProxy();
+        // ...
+
+        if (config.env?.schema) {
+          overrideProcessEnv({
+            getEnv,
+            variables: Object.keys(config.env.schema).map((destKey) => ({
+              destKey,
+            })),
+          });
+        }
+      },
+    },
+  };
+};
+```
+
+##### Inlining build only variables
+
+`overrideProcessEnv` also allows to provide some build only variables for runtime, eg. [Netlify Build environment variables](https://docs.netlify.com/configure-builds/environment-variables/).
+
+An adapter can provide its own variables and override `process.env` accordingly. For example:
+
+```ts
+import { overrideProcessEnv } from "astro/runtime/server/astro-env.js"
+import { getEnv } from "./utils.js"
+
+const integration = (): AstroIntegration => {
+  let config;
+
+  return {
+    name: "adapter",
+    hooks: {
+      "astro:config:setup": ({ updateConfig }) => {
+        updateConfig({
+          env: {
+            schema: {
+              PUBLIC_NETLIFY: envField.boolean({
+								context: 'server',
+								access: 'public',
+								default: false,
+							}),
+							PUBLIC_DEPLOY_URL: envField.string({
+								context: 'server',
+								access: 'public',
+								optional: true,
+							}),
+            }
+          }
+        })
+      }
+      "astro:config:done": () => {
+        overrideProcessEnv({
+					getEnv,
+					variables: [
+						{
+							destKey: 'PUBLIC_NETLIFY',
+							srcKey: 'NETLIFY',
+						},
+						{
+							destKey: 'PUBLIC_DEPLOY_URL',
+							srcKey: 'DEPLOY_URL',
+							default: url,
+						},
+					],
+				});
+			}
```

Comment:
I don't think an adapter developer should import `overrideProcessEnv` like this (from the runtime, using a long import). Instead, I think we should provide this function **via hook**, maybe in `astro:config:done` (or another hook before the build). As you explained in the RFC, this is required during the dev/build, and we want to have these environment variables available **before** generating and rendering an actual page.

This way, we can control it, and we will provide this function **only if the feature is enabled by the user and supported by the adapter**. Win-win.

We could call it `setOverrideProcessEnv`. 

Also, can you explain what `default`, `destKey`, and `srcKey` are? I looked at the source code and adapter PR, and I couldn't understand them.

Reply:
> * It can be called in quite a few hooks, from `astro:config:setup` to `astro:build:setup`

Maybe I worded my thoughts badly, sorry! I meant, let's choose only **one** hook. If it's required for dev too, maybe `astro:config:setup` makes more sense.

> * It does nothing that involves astro (eg. settings)

Why? It sets an internal runtime function that is used inside the rendering phase (components that can read the environment variables)



> * `destKey` is the key that should be set on `process.env`
> 
> * `srcKey` is the key passed to `getEnv` inside the function. For example on netlify, that's important because `NETLIFY` is exposed but it has to be `PUBLIC_NETLIFY` with Astro Env, so it helps reducing boilerplate
> 
> * `default` is meant to be used as a fallback value, for example for `DEPLOY_URL` in dev

Awesome! I am not very good at naming, so I am happy with the one chosen. It would be great if we could document them

Reply:
> Why? It sets an internal runtime function that is used inside the rendering phase (components that can read the environment variables)

It's not! It's just setting `process.env` so it doesn't call any internal astro stuff.

> Maybe I worded my thoughts badly, sorry! I meant, let's choose only one hook. If it's required for dev too, maybe astro:config:setup makes more sense.

Well it's not exactly true. Yes `astro:config:setup` makes sense for dev but I would also expose it another hook that runs in both dev and build (probably `astro:config:done`) to allow the netlify build env variable inlining

Reply:
If all this does is set a variable on `process.env` then why do we need an API for this at all? Why don't they just set the variable themselves.

```js
"astro:config:done": () => {
  process.env.PUBLIC_NETLIFY = process.env.NETLIFY;
}
```

Reply:
Given this is a large enough RFC already, I think adding this extra complexity (and maintenance) is not worth it at this time. I don't think this adds enough value to justify the maintenance in Astro. Integrations are advanced usage so they should be able to manage conditionally setting up environment variables in dev, if needed.